File: sender.js
/**
 * sender.js — sends the digest to a Telegram chat via Bot API.
 * Splits messages that exceed Telegram's 4096-char limit.
 */

import { BOT_TOKEN, TARGET_CHAT_ID } from "./config.js";

const TG_API  = `https://api.telegram.org/bot${BOT_TOKEN}`;
const MAX_LEN = 4096;
const RETRIES = 3;
const RETRY_DELAY = 5000;

function sleep(ms) {
  return new Promise((r) => setTimeout(r, ms));
}

function splitMessage(text, maxLen = MAX_LEN) {
  if (text.length <= maxLen) return [text];

  const chunks = [];
  let current  = "";

  for (const block of text.split("\n\n")) {
    const candidate = current ? current + "\n\n" + block : block;
    if (candidate.length > maxLen) {
      if (current) chunks.push(current.trim());
      current = block;
    } else {
      current = candidate;
    }
  }
  if (current.trim()) chunks.push(current.trim());
  return chunks;
}

async function postPart(text, attempt = 1) {
  const res = await fetch(`${TG_API}/sendMessage`, {
    method:  "POST",
    headers: { "Content-Type": "application/json" },
    body:    JSON.stringify({
      chat_id:                  TARGET_CHAT_ID,
      text,
      parse_mode:               "HTML",
      disable_web_page_preview: true,
    }),
  });

  const data = await res.json();
  if (data.ok) return;

  if (attempt < RETRIES) {
    console.warn(`[sender] Retry ${attempt}: ${JSON.stringify(data)}`);
    await sleep(RETRY_DELAY);
    return postPart(text, attempt + 1);
  }

  throw new Error(`Telegram API error: ${JSON.stringify(data)}`);
}

/**
 * @param {string} title  — header line
 * @param {string} digest — HTML body
 */
export async function sendDigest(title, digest) {
  const full   = `${title}\n\n${digest}`;
  const parts  = splitMessage(full);

  console.log(`[sender] Sending digest (${parts.length} part(s)) …`);

  for (let i = 0; i < parts.length; i++) {
    const suffix = parts.length > 1 ? `\n\n<i>(${i + 1}/${parts.length})</i>` : "";
    await postPart(parts[i] + suffix);
    if (i < parts.length - 1) await sleep(1000); // flood-limit guard
  }

  console.log("[sender] Digest sent ✓");
}

File: sender.py
"""
sender.py — sends the final digest to a Telegram channel/chat via Bot API.
Splits messages that exceed Telegram's 4096-character limit.
"""

from __future__ import annotations

import asyncio
import logging
import httpx

from config import BOT_TOKEN, TARGET_CHAT_ID

log = logging.getLogger(__name__)

_TG_API = f"https://api.telegram.org/bot{BOT_TOKEN}"
_MAX_LEN = 4096
_RETRY_ATTEMPTS = 3
_RETRY_DELAY = 5   # seconds


def _split_message(text: str, max_len: int = _MAX_LEN) -> list[str]:
    """Split on blank lines to keep topics together; never cut mid-paragraph."""
    if len(text) <= max_len:
        return [text]

    chunks: list[str] = []
    current = ""
    for block in text.split("\n\n"):
        candidate = (current + "\n\n" + block).lstrip()
        if len(candidate) > max_len:
            if current:
                chunks.append(current.strip())
            current = block
        else:
            current = candidate
    if current.strip():
        chunks.append(current.strip())
    return chunks


async def _post(client: httpx.AsyncClient, text: str) -> None:
    payload = {
        "chat_id": TARGET_CHAT_ID,
        "text": text,
        "parse_mode": "HTML",
        "disable_web_page_preview": True,
    }
    for attempt in range(1, _RETRY_ATTEMPTS + 1):
        try:
            resp = await client.post(f"{_TG_API}/sendMessage", json=payload, timeout=30)
            data = resp.json()
            if data.get("ok"):
                return
            log.error(f"Telegram API error: {data}")
        except httpx.RequestError as exc:
            log.warning(f"Request error (attempt {attempt}): {exc}")
        if attempt < _RETRY_ATTEMPTS:
            await asyncio.sleep(_RETRY_DELAY)
    raise RuntimeError("All retry attempts to Telegram API exhausted.")


async def send_digest(title: str, digest: str) -> None:
    """Send title + digest (possibly multi-part) to TARGET_CHAT_ID."""
    full_text = f"{title}\n\n{digest}"
    parts = _split_message(full_text)

    async with httpx.AsyncClient() as client:
        for i, part in enumerate(parts, 1):
            suffix = f"\n\n<i>({i}/{len(parts)})</i>" if len(parts) > 1 else ""
            await _post(client, part + suffix)
            if len(parts) > 1:
                await asyncio.sleep(1)   # avoid flood limits

    log.info(f"Digest sent ({len(parts)} part(s)).")
